### ISSUE TITLE
banning not working 

### ISSUE THREAD

**Issue by David1ali12** [Barbosik/MultiOgar#228]
when I ban ppl  I just see them simply respawn

**Comment by Barbosik** [comment]
probably you're ban wrong people, for example several people may have the same name :)

**Comment by NatsuTheGreat** [comment]
It is very easy to get through this banning system, a simple extension can do it.

**Comment by David1ali12** [comment]
@Barbosik  no there about 10-20 players in my servers and when I banned some of them they just rejoin normally 
they are not using vpn doe maybe the problem is with my MultiOgar version I'm using older version.

**Comment by David1ali12** [comment]
I have another problem I dont want to make another issue for that I was adding a new command for my server in command list it was called troll command works fine but does not change player color this my code can any1 explain what I did wrong thanks 

color does not change to black :/
```
 troll: function(gameServer, split) {

       
 var id = parseInt(split[1]);

        if (isNaN(id)) {

            Logger.warn("Please specify a valid player ID!");

            return;

        }

        var size = Math.sqrt(200 * 100);



        // Sets mass to the specified amount

        for (var i in gameServer.clients) {

            if (gameServer.clients[i].playerTracker.pID == id) {

                var client3 = gameServer.clients[i].playerTracker;

                for (var j in client3.cells) {

                    client3.cells[j].setSize(size);

                } 
            }
        }
        
    


        var name = "GotTrolled:;-;";




        // Change name

        for (var i = 0; i < gameServer.clients.length; i++) {

            var client = gameServer.clients[i].playerTracker;



            if (client.pID == id) {

                console.log("Changing " + client.getFriendlyName() + " to " + name);

                client.setName(name);

                return;

            }

        }



        for (var i in gameServer.clients) {
    if (gameServer.clients[i].playerTracker.pID == id) {
      var client4 = gameServer.clients[i].playerTracker;
      client4.setColor(0); // Set color
      for (var j in client4.cells) {
        client4.cells[j].setColor(0);
      }

                return;

            }

        }
        // Error

        Logger.warn("Player " + id + " was not found");

    },
```

**Comment by NatsuTheGreat** [comment]
Can you please try:

```
 troll: function(gameServer, split) {


     var id = parseInt(split[1]);

     if (isNaN(id)) {

         Logger.warn("Please specify a valid player ID!");

         return;

     }

     var size = Math.sqrt(200 * 100);



     // Sets mass to the specified amount

     for (var i in gameServer.clients) {

         if (gameServer.clients[i].playerTracker.pID == id) {

             var client3 = gameServer.clients[i].playerTracker;

             for (var j in client3.cells) {

                 client3.cells[j].setSize(size);

             }
         }
     }




     var name = "GotTrolled:;-;";




     // Change name

     for (var i = 0; i < gameServer.clients.length; i++) {

         var client = gameServer.clients[i].playerTracker;



         if (client.pID == id) {

             console.log("Changing " + client.getFriendlyName() + " to " + name);

             client.setName(name);

             return;

         }

     }



     for (var i in gameServer.clients) {
         if (gameServer.clients[i].playerTracker.pID == id) {
             var client4 = gameServer.clients[i].playerTracker;
             for (var j = 0; j < client4.cells.length; j++) {
                 client4.cells[j].setColor({'r': 0, 'g': 0, 'b': 0 });
             }

             return;

         }

     }
     // Error

     Logger.warn("Player " + id + " was not found");

 },
```

**Comment by Barbosik** [comment]
@David1ali12: solution from @NatsuTheGreat is good. It will works. But there is also need to add a small, but important fix:
```
client3.cells[j].setSize(size);
gameServer.updateNodeQuad(client3.cells[j]);
```

**Comment by Barbosik** [comment]
@David1ali12: if you trying to ban bots, which connected from the same IP where your server runs, the ban will not work. Because server cannot ban self IP.

**Comment by David1ali12** [comment]
lol not bots I mean I was trying to ban real players :/ can you make system better banning this one does not work fine

**Comment by Barbosik** [comment]
@David1ali12: are you using ws or uws?

name will not be updated immediately, the player will notice name change only after split and rejoin. Other players will notice name change after some time. Color should be changed immediately.

**Comment by NatsuTheGreat** [comment]
Color is always going to change, because it always updates in UpdateNodes. Both in updNodes and addNodes. So it always checks for color. And if the still doesn't change i'll look into it now.

**Comment by David1ali12** [comment]
@Barbosik I'm using uws 
@NatsuTheGreat  can you test it does n't work for me

**Comment by Barbosik** [comment]
@David1ali12: it seems like uws doesn't provide information about remote IP address. Check your CONNECT/DISCONNECT records in the log file. Does it consists valid IP address?

**Comment by NatsuTheGreat** [comment]
Mine consists of ::ff:ff:f: and such and then with the IP Address.

**Comment by Barbosik** [comment]
@NatsuTheGreat: it means that there is a problem with network interface binding in your uws implementation

**Comment by NatsuTheGreat** [comment]
:/

**Comment by David1ali12** [comment]
I tested with ws and uws does not matter and I banned my self I could simply rejoin it just kicked me

**Comment by David1ali12** [comment]
```
[NONE ][14:33:32]  ID     | IP              | P | NICK            | CELLS | SCORE  | POSITION    
[NONE ][14:33:32] -------------------------------------------------------------------------------
[NONE ][14:33:32]       1 | 10.240.1.13     | 5 | ;-;             |     1 |   7311 | -3713, -2306
[NONE ][14:33:32]       3 | 10.240.0.224    | 5 | ;-;             |     1 |  13185 |  3920, -7013
[NONE ][14:33:57] The IP 10.240.0.224 has been banned
[NONE ][14:33:57] Banned: ";-;" with Player ID 3
[NONE ][14:33:57] DISCONNECTED 10.240.0.224:34898, code: 1000, reason: "Banned from server", name: ";-;"
[INFO ][14:33:57] 2 IP ban records saved.
[NONE ][14:34:10] CONNECTED    10.240.0.215:59662, origin: "http://agarlist.com"
[INFO ][14:35:21] Showing 2 players: 
then  I could rejoin 
[NONE ][14:35:21]  ID     | IP              | P | NICK            | CELLS | SCORE  | POSITION    
[NONE ][14:35:21] -------------------------------------------------------------------------------
[NONE ][14:35:21]       1 | 10.240.1.13     | 5 | ;-;             |     1 |   5958 | -3704, -2347
[NONE ][14:35:21]       4 | 10.240.0.215    | 5 | ;-;             |     1 |  23061 | -3296, -6878


```

**Comment by NatsuTheGreat** [comment]
My banning works, but this.emit() gives a problem :/

**Comment by Barbosik** [comment]
I will try to test uws support soon. Worked on ogar-tracker.tk and server optimizations and have no time

**Comment by David1ali12** [comment]
but its with ws now :/

**Comment by NatsuTheGreat** [comment]
I finished David. I found out that the problem was with an incorrect ```return;``` statement that was stated before the change color function. But I couldn't bear to make the function shorter, I am like Barbosik, I like short code. :smile: Here it is:

```
	 troll: function(gameServer, split) {
	     var id = parseInt(split[1]);

	     if (isNaN(id)) {
	         Logger.warn("Please specify a valid player ID!");
	         return;
	     }

	     var size = Math.sqrt(200 * 100);
	     var name = "GotTrolled:;-;";

	     for (var i in gameServer.clients) {
	         if (gameServer.clients[i].playerTracker.pID == id) {
	             var client = gameServer.clients[i].playerTracker;
	             for (var j in client.cells) {
	                 client.cells[j].setSize(size); // Sets the size.
	                 if (client.pID == id) { // Changes name.
	                     console.log("[Console] Changing " + client.getFriendlyName() + " to " + name);
	                     client.setName(name);
	                 }
	                 client.cells[j].setColor({
	                     'r': 0,
	                     'g': 0,
	                     'b': 0
	                 }); // Changes color.
	                 return;
	             }
	         }
	     }
	     // Error
	     Logger.warn("Player " + id + " was not found");
	 },
```

**Comment by Barbosik** [comment]
@David1ali12: he just using another IP :) You can use subnetwork ban for such cases:
```
ban 10.240.*.*
```

I'm planning to add standard IP mask soon, but currently you can use * instead of number in IP.

**Comment by NatsuTheGreat** [comment]
I made a little mistake, reload David. I edited my comment.

**Comment by David1ali12** [comment]
@Barbosik  thx should this be closed 
@NatsuTheGreat  thx mate :+1:

**Comment by NatsuTheGreat** [comment]
@David1ali12 Have fun with it. xP
